yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -466,6 +466,146 @@ class FromPublisherOperator : public Flowable<T> {
+  }
+
+ private:
+  using SuperSubscription = typename Super::Subscription;
+  class FMSubscription : public SuperSubscription {
+   public:
+    FMSubscription(
+        Reference<ThisOperatorT> flowable,
+        Reference<Subscriber<R>> subscriber)
+        : SuperSubscription(std::move(flowable), std::move(subscriber)) {}
+
+    void onNextImpl(T value) override {
+      LOG(INFO) << "FMSubscription::onNextImpl(" << value << ")";
+      CHECK(!currentStreamSubscriber_ && pending_);
+
+      auto&& flatMapOp = this->getFlowableOperator();
+      Reference<Flowable<R>> newStream = flatMapOp->function_(std::move(value));
+      currentStreamSubscriber_ =
+          yarpl::make_ref<MappedStreamSubscriber>(this->ref_from_this(this));
+      newStream->subscribe(currentStreamSubscriber_);
+    }
+
+    void onCompleteImpl() override {
+      LOG(INFO) << "FMSubscription::onCompleteImpl()";
+      SuperSubscription::onCompleteImpl();
+    }
+
+    void onErrorImpl(folly::exception_wrapper ex) override {
+      LOG(INFO) << "FMSubscription::onErrorImpl()";
+      SuperSubscription::onErrorImpl(std::move(ex));
+    }
+
+    // called from MappedStreamSubscriber
+    void mappedOnNextImpl(R value) {
+      LOG(INFO) << "FMSubscription::mappedOnNextImpl(" << value << ")";
+      requested_--;
+      SuperSubscription::subscriberOnNext(std::move(value));
+    }
+
+    void mappedOnCompleteImpl() {
+      LOG(INFO) << "FMSubscription::mappedOnCompleteImpl()";
+
+      CHECK(!currentStreamSubscriber_);
+      CHECK(!pending_);
+
+      if (requested_) {
+        pending_ = true;
+        SuperSubscription::request(1);
+      }
+    }
+
+    void request(int64_t n) override {
+      LOG(INFO) << "FMSubscription::request(" << n << ")";
+      requested_ += n;
+
+      if (!currentStreamSubscriber_ && !pending_) {
+        LOG(INFO) << "FMSubscription::request(" << n
+                  << ") nothing pending, requesting!";
+        pending_ = true;
+        SuperSubscription::request(1);
+      } else {
+        CHECK(currentStreamSubscriber_);
+        LOG(INFO) << "FMSubscription::request(" << n
+                  << ") passing request through to currentStreamSubscriber_";
+        currentStreamSubscriber_->request(n);
+      }
+    }
+
+    void cancel() override {
+      LOG(INFO) << "FMSubscription::cancel()";
+      SuperSubscription::cancel();
+    }
+
+   private:
+    class MappedStreamSubscriber : public BaseSubscriber<R> {
+     public:
+      MappedStreamSubscriber(Reference<FMSubscription> subscription)
+          : flatMapSubscription_(std::move(subscription)) {}
+
+      void onSubscribeImpl() final {
+        LOG(INFO) << "MappedStreamSubscriber::onSubscribeImpl()";
+        if (flatMapSubscription_->requested_) {
+          LOG(INFO) << "MappedStreamSubscriber::onSubscribeImpl(): requsting "
+                    << flatMapSubscription_->requested_;
+          BaseSubscriber<R>::request(flatMapSubscription_->requested_);
```

Comment:
as mentioned above, you can't give all credits just to one subscriptions because the other ones would never make any progress.

Reply:
n/a with updates